### PR TITLE
Add mellanox plugin

### PR DESCRIPTION
**Pull request by Mmduh-483**
Change Mlnx devices link type to ETH if it is another type, change the total VFs to the requested num of VFs, disable SRIOV for Mlnx cards if the num of VFs is 0 otherwise enable it.

### CODE REVIEW

File: pkg/apis/sriovnetwork/v1/helper.go [hunk 1]
```diff
@@ -84,9 +84,12 @@ func (p *SriovNetworkNodePolicy) Apply(state *SriovNetworkNodeState) {
...
 			interfaces = append(interfaces, Interface{
 				PciAddress: iface.PciAddress,
-				Mtu:        p.Spec.Mtu,
-				NumVfs:     p.Spec.NumVfs,
-				DeviceType: p.Spec.DeviceType,
+				Mtu:         p.Spec.Mtu,
+				NumVfs:      p.Spec.NumVfs,
+				TotalVfs:    p.Spec.TotalVfs,
+				LinkType:    p.Spec.LinkType,
+				EnableSriov: p.Spec.EnableSriov,
```
Comment (moshe010) [inline pkg/apis/sriovnetwork/v1/helper.go]: you don't need the TotalVfs LinkType EnableSriov. we just need to make sure that if you have NumVfs is should be equal to the total number of VF.  EnableSriov should be set to true when you have NumVfs >0.   regarding the LinkType we can now assume it only ethernet

File: pkg/apis/sriovnetwork/v1/sriovnetworknodepolicy_types.go [hunk 2]
```diff
@@ -21,6 +21,9 @@ type SriovNetworkNodePolicySpec struct {
 	NicSelector  SriovNetworkNicSelector `json:"nicSelector"`
 	DeviceType   string                  `json:"deviceType,omitempty"`
 	IsRdma       bool                    `json:"isRdma,omitempty"`
+	TotalVfs     int                     `json:"totalVfs,omitempty"`
+	LinkType     string                  `json:"linkType,omitempty"` 
+	EnableSriov  bool                    `json:"enableSriov,omitempty"`
```
Comment (moshe010) [inline pkg/apis/sriovnetwork/v1/sriovnetworknodepolicy_types.go]: same comment as above

File: pkg/daemon/daemon.go [hunk 3]
```diff
@@ -410,6 +410,7 @@ func needRestartDevicePlugin(oldState, newState *sriovnetworkv1.SriovNetworkNode
 func registerPlugins(ns *sriovnetworkv1.SriovNetworkNodeState) []string {
 	pluginNames := make(map[string]bool)
 	for _, iface := range ns.Status.Interfaces {
+		glog.Infof("addr: %s , vendor: %s",iface.PciAddress , iface.Vendor)
```
Comment (moshe010) [inline pkg/daemon/daemon.go]: seem like unrelated change

File: pkg/plugins/mellanox/mellanox_plugin.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	return
+}
+
+// Apply config change
+func (p *MellanoxPlugin) Apply() error {
+	glog.Info("mellanox-plugin Apply()")
+	return nil
+}
+
+func configIface(pciAddress string, args []string) error {
```
Comment (moshe010) [inline pkg/plugins/mellanox/mellanox_plugin.go]: I think configFW is better

File: pkg/plugins/mellanox/mellanox_plugin.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+				mstconfigArgs = append(mstconfigArgs, fmt.Sprintf("NUM_OF_VFS=%s", ifaceSpec.TotalVfs))
+			}
+			if (iface.EnableSriov != ifaceSpec.EnableSriov) {
+				mstconfigArgs = append(mstconfigArgs, fmt.Sprintf("SRIOV_EN=%s", ifaceSpec.EnableSriov))
+			}
+			if (iface.LinkType != ifaceSpec.LinkType) {
+//				mstconfigArgs = append(mstconfig, "LINK_TYPE" + ifaceSpec.LinkType)
+			}
+			if (len(mstconfigArgs) > 0) {
+				configIface(iface.PciAddress, mstconfigArgs)
```
Comment (moshe010) [inline pkg/plugins/mellanox/mellanox_plugin.go]: this should be done in the apply

File: pkg/plugins/mellanox/mellanox_plugin.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+			if (iface.TotalVfs != ifaceSpec.TotalVfs) {
+				mstconfigArgs = append(mstconfigArgs, fmt.Sprintf("NUM_OF_VFS=%s", ifaceSpec.TotalVfs))
+			}
+			if (iface.EnableSriov != ifaceSpec.EnableSriov) {
+				mstconfigArgs = append(mstconfigArgs, fmt.Sprintf("SRIOV_EN=%s", ifaceSpec.EnableSriov))
+			}
+			if (iface.LinkType != ifaceSpec.LinkType) {
+//				mstconfigArgs = append(mstconfig, "LINK_TYPE" + ifaceSpec.LinkType)
+			}
+			if (len(mstconfigArgs) > 0) {
```
Comment (moshe010) [inline pkg/plugins/mellanox/mellanox_plugin.go]: you should query mstconfig to get the values and set them according to NumVfs.
MAX_NUM_OF should be == NumVfs.
LinkType == Ethernet (fror now)
EnableSriov == Yes all the time.

File: pkg/utils/utils.go [hunk 7]
```diff
@@ -347,3 +361,64 @@ func Chroot(path string) (func() error, error) {
 		return syscall.Chroot(".")
 	}, nil
 }
+
+func MlnxIfaceLinkType(pciAddress string) (string, error) {
```
Comment (moshe010) [inline pkg/utils/utils.go]: this should be moved to Mellanox plugin

File: pkg/apis/sriovnetwork/v1/sriovnetworknodepolicy_types.go (hunk 2 above)
Comment (pliurh) [inline pkg/apis/sriovnetwork/v1/sriovnetworknodepolicy_types.go]: I don't think we need to add `EnableSriov` field explicitly here. When a user specified `NumVfs`, it implies that SR-IOV shall be enable.

File: pkg/apis/sriovnetwork/v1/sriovnetworknodepolicy_types.go [hunk 8]
```diff
@@ -21,6 +21,9 @@ type SriovNetworkNodePolicySpec struct {
 	NicSelector  SriovNetworkNicSelector `json:"nicSelector"`
 	DeviceType   string                  `json:"deviceType,omitempty"`
 	IsRdma       bool                    `json:"isRdma,omitempty"`
+	TotalVfs     int                     `json:"totalVfs,omitempty"`
+	LinkType     string                  `json:"linkType,omitempty"`
```
Comment (pliurh) [inline pkg/apis/sriovnetwork/v1/sriovnetworknodepolicy_types.go]: This `LinkType` is only supported by MLX nics, right? I am not quite sure if we shall add it. As you know, we are not going to support infiniband in this release, so I don't think we should expose this mode select option, at least for now. @zshi-redhat what do you think?

File: pkg/apis/sriovnetwork/v1/sriovnetworknodepolicy_types.go [hunk 9]
```diff
@@ -21,6 +21,9 @@ type SriovNetworkNodePolicySpec struct {
 	NicSelector  SriovNetworkNicSelector `json:"nicSelector"`
 	DeviceType   string                  `json:"deviceType,omitempty"`
 	IsRdma       bool                    `json:"isRdma,omitempty"`
+	TotalVfs     int                     `json:"totalVfs,omitempty"`
```
Comment (pliurh) [inline pkg/apis/sriovnetwork/v1/sriovnetworknodepolicy_types.go]: The `TotalVfs` in SriovNetworkNodePolicy may looks confusing to users, since not all the NICs support configuring this parameter. And I think, from function perspective, specifying `NumVfs` is enough.

File: pkg/plugins/mellanox/mellanox_plugin.go [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+// OnNodeStateAdd Invoked when SriovNetworkNodeState CR is created, return if need dain and/or reboot node
+func (p *MellanoxPlugin) OnNodeStateAdd(state *sriovnetworkv1.SriovNetworkNodeState) (needDrain bool, needReboot bool, err error) {
+	glog.Info("mellanox-plugin OnNodeStateAdd()")
+
+	return p.OnNodeStateChange(nil, state)
+}
+
+// OnNodeStateChange Invoked when SriovNetworkNodeState CR is updated, return if need dain and/or reboot node
+func (p *MellanoxPlugin) OnNodeStateChange(old, new *sriovnetworkv1.SriovNetworkNodeState) (needDrain bool, needReboot bool, err error) {
+	glog.Info("mellanoxi-Plugin OnNodeStateChange()")
```
Comment (moshe010) [inline pkg/plugins/mellanox/mellanox_plugin.go]: s/mellanoxi/mellanox

File: pkg/plugins/mellanox/mellanox_plugin.go [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+	}
+	glog.Infof("mellanox needDrain %v needReboot %v",needDrain,needReboot)
+	return
+}
+
+// Apply config change
+func (p *MellanoxPlugin) Apply() error {
+	glog.Info("mellanox-plugin Apply()")
+	err := configFW()
+	if err != nil {
```
Comment (moshe010) [inline pkg/plugins/mellanox/mellanox_plugin.go]: just do return configFW()

File: pkg/plugins/mellanox/mellanox_plugin.go [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+		if requireChange {
+			attributesToChange[iface.PciAddress] = *attrs
+			needReboot = true
+		}
+	}
+
+	if needReboot {
+		needDrain = true
+	}
+	glog.Infof("mellanox needDrain %v needReboot %v",needDrain,needReboot)
```
Comment (moshe010) [inline pkg/plugins/mellanox/mellanox_plugin.go]: add spaces ("mellanox needDrain %v needReboot %v", needDrain, needReboot)

File: pkg/plugins/mellanox/mellanox_plugin.go [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+	for _, iface := range new.Spec.Interfaces {
+		if !isMlnxNic(iface.PciAddress, new) {
+			continue
+		}
+		ifaceData, err := getMlnxNicData(iface.PciAddress)
+		if err != nil {
+			return false, false, err
+		}
+		attrs := &fwAttributes{totalVfs: -1}
+		requireChange := false
```
Comment (moshe010) [inline pkg/plugins/mellanox/mellanox_plugin.go]: where requireChange is set to true?

File: pkg/plugins/mellanox/mellanox_plugin.go [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func isMlnxNic(pciAddress string, state *sriovnetworkv1.SriovNetworkNodeState) bool {
+	for _, iface := range state.Status.Interfaces {
+		if iface.PciAddress == pciAddress {
+			if iface.Vendor == "15b3" {
```
Comment (moshe010) [inline pkg/plugins/mellanox/mellanox_plugin.go]: 15b3 move to constant

File: pkg/daemon/daemon.go [hunk 15]
```diff
@@ -250,7 +250,7 @@ func (dn *Daemon) nodeStateChangeHandler(old, new interface{}) {
 		}
 	}
 
-	if len(dn.LoadedPlugins) > 1 {
+	if len(dn.LoadedPlugins) > 1 && !reqReboot {
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/daemon/daemon.go]: @pliurh is it ok to see this failure with Dockerfile.rhel7 and Dockerfile.sriov-network-config-daemon.rhel7
```
Step 7/16 : RUN yum -y update; yum -y install mstflint; yum clean all                                            
 ---> Running in 10d6f84d73aa                                                                                    
Loaded plugins: ovl, product-id, search-disabled-repos, subscription-manager                                     
This system is not registered with an entitlement server. You can use subscription-manager to register.          
http://base-4-0.ocp.svc/rhel-fast-datapath/repodata/repomd.xml: [Errno 14] curl#6 - "Could not resolve host: base-4-0.ocp.svc; Unknown error"
Trying other mirror.                                                                                                                                                                                          
                                                                                                                                                                                                              
                                                                                                                                                                                                              
 One of the configured repositories failed (rhel-fast-datapath),                                                                                                                                              
 and yum doesn't have enough cached data to continue. At this point the only                                                                                                                                  
 safe thing yum can do is fail. There are a few ways to work "fix" this:                                                                                                                                      
                                                                                                                                                                                                              
     1. Contact the upstream for the repository and get them to fix the problem.                                                                                                                              
                                                                                                                                                                                                              
     2. Reconfigure the baseurl/etc. for the repository, to point to a working                                                                                                                                
        upstream. This is most often useful if you are using a newer                                                                                                                                          
        distribution release than is supported by the repository (and the                                                                                                                                     
        packages for the previous distribution release still work).                                                                                                                                           
                                                                                                                                                                                                              
     3. Run the command with the repository temporarily disabled                                                                                                                                              
            yum --disablerepo=rhel-fast-datapath ...                                                                                                                                                          
                                                                                                                                                                                                              
     4. Disable the repository permanently, so yum won't use it by default. Yum                                                                                                                               
        will then just ignore the repository until you permanently enable it                                                                                                                                  
        again or use --enablerepo for temporary usage:                                                                                                                                                        
                                                                                                                                                                                                              
            yum-config-manager --disable rhel-fast-datapath                                                                                                                                                   
        or                                                                                                                                                                                                    
            subscription-manager repos --disable=rhel-fast-datapath                                                                                                                                           
                                                                                                                                                                                                              
     5. Configure the failing repository to be skipped, if it is unavailable.                                                                                                                                 
        Note that yum will try to contact the repo. when it runs most commands,                                                                                                                               
        so will have to try and fail each time (and thus. yum will be be much                                                                                                                                 
        slower). If it is a very temporary problem though, this is often a nice                                                                                                                               
        compromise:                                                                                                                                                                                           
                                                                                                                                                                                                              
            yum-config-manager --save --setopt=rhel-fast-datapath.skip_if_unavailable=true                                                                                                                    
                                                                                                                                                                                                              
failure: repodata/repomd.xml from rhel-fast-datapath: [Errno 256] No more mirrors to try.                                                                                                                     
http://base-4-0.ocp.svc/rhel-fast-datapath/repodata/repomd.xml: [Errno 14] curl#6 - "Could not resolve host: base-4-0.ocp.svc; Unknown error"                                                                 
Loaded plugins: ovl, product-id, search-disabled-repos, subscription-manager                                                                                                                                  
This system is not registered with an entitlement server. You can use subscription-manager to register.                                                                                                       
http://base-4-0.ocp.svc/rhel-fast-datapath/repodata/repomd.xml: [Errno 14] curl#6 - "Could not resolve host: base-4-0.ocp.svc; Unknown error"                                                                 
Trying other mirror.                                                                                                                                                                                          
                                                                                                                                                                                                              
                                                                                                                                                                                                              
 One of the configured repositories failed (rhel-fast-datapath),                                                                                                                                              
 and yum doesn't have enough cached data to continue. At this point the only                                                                                                                                  
 safe thing yum can do is fail. There are a few ways to work "fix" this:                                                                                                                                      
                                                                                                                                                                                                              
     1. Contact the upstream for the repository and get them to fix the problem.                                                                                                                              
                                                                                                                                                                                                              
     2. Reconfigure the baseurl/etc. for the repository, to point to a working                                                                                                                                
        upstream. This is most often useful if you are using a newer                                                                                                                                          
        distribution release than is supported by the repository (and the                                                                                                                                     
        packages for the previous distribution release still work).                                                                                                                                           
                                                                                                                                                                                                              
     3. Run the command with the repository temporarily disabled                                                                                                                                              
            yum --disablerepo=rhel-fast-datapath ...                                                                                                                                                          
                                                                                                                                                                                                              
     4. Disable the repository permanently, so yum won't use it by default. Yum                                                                                                                               
        will then just ignore the repository until you permanently enable it                                                                                                                                  
        again or use --enablerepo for temporary usage:                                                                                                                                                        
                                                                                                                                                                                                              
            yum-config-manager --disable rhel-fast-datapath                                                                                                                                                   
        or                                                                                                                                                                                                    
            subscription-manager repos --disable=rhel-fast-datapath                                                                                                                                           

     5. Configure the failing repository to be skipped, if it is unavailable.
        Note that yum will try to contact the repo. when it runs most commands,
        so will have to try and fail each time (and thus. yum will be be much
        slower). If it is a very temporary problem though, this is often a nice
        compromise:

            yum-config-manager --save --setopt=rhel-fast-datapath.skip_if_unavailable=true

failure: repodata/repomd.xml from rhel-fast-datapath: [Errno 256] No more mirrors to try.
http://base-4-0.ocp.svc/rhel-fast-datapath/repodata/repomd.xml: [Errno 14] curl#6 - "Could not resolve host: base-4-0.ocp.svc; Unknown error"
Loaded plugins: ovl, product-id, search-disabled-repos, subscription-manager
This system is not registered with an entitlement server. You can use subscription-manager to register.
Cleaning repos: rhel-fast-datapath rhel-server rhel-server-extras
              : rhel-server-optional rhel-server-ose rhel-server-rhscl
Removing intermediate container 10d6f84d73aa

```

File: Dockerfile.sriov-network-config-daemon [hunk 16]
```diff
@@ -5,6 +5,7 @@ RUN make _build-sriov-network-config-daemon
 RUN make plugins
 
 FROM centos:7
+RUN yum install -y mstflint
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply Dockerfile.sriov-network-config-daemon]: @pliurh is it ok to see this failure with Dockerfile.rhel7 and Dockerfile.sriov-network-config-daemon.rhel7
```
Step 7/16 : RUN yum -y update; yum -y install mstflint; yum clean all                                            
 ---> Running in 10d6f84d73aa                                                                                    
Loaded plugins: ovl, product-id, search-disabled-repos, subscription-manager                                     
This system is not registered with an entitlement server. You can use subscription-manager to register.          
http://base-4-0.ocp.svc/rhel-fast-datapath/repodata/repomd.xml: [Errno 14] curl#6 - "Could not resolve host: base-4-0.ocp.svc; Unknown error"
Trying other mirror.                                                                                                                                                                                          
                                                                                                                                                                                                              
                                                                                                                                                                                                              
 One of the configured repositories failed (rhel-fast-datapath),                                                                                                                                              
 and yum doesn't have enough cached data to continue. At this point the only                                                                                                                                  
 safe thing yum can do is fail. There are a few ways to work "fix" this:                                                                                                                                      
                                                                                                                                                                                                              
     1. Contact the upstream for the repository and get them to fix the problem.                                                                                                                              
                                                                                                                                                                                                              
     2. Reconfigure the baseurl/etc. for the repository, to point to a working                                                                                                                                
        upstream. This is most often useful if you are using a newer                                                                                                                                          
        distribution release than is supported by the repository (and the                                                                                                                                     
        packages for the previous distribution release still work).                                                                                                                                           
                                                                                                                                                                                                              
     3. Run the command with the repository temporarily disabled                                                                                                                                              
            yum --disablerepo=rhel-fast-datapath ...                                                                                                                                                          
                                                                                                                                                                                                              
     4. Disable the repository permanently, so yum won't use it by default. Yum                                                                                                                               
        will then just ignore the repository until you permanently enable it                                                                                                                                  
        again or use --enablerepo for temporary usage:                                                                                                                                                        
                                                                                                                                                                                                              
            yum-config-manager --disable rhel-fast-datapath                                                                                                                                                   
        or                                                                                                                                                                                                    
            subscription-manager repos --disable=rhel-fast-datapath                                                                                                                                           
                                                                                                                                                                                                              
     5. Configure the failing repository to be skipped, if it is unavailable.                                                                                                                                 
        Note that yum will try to contact the repo. when it runs most commands,                                                                                                                               
        so will have to try and fail each time (and thus. yum will be be much                                                                                                                                 
        slower). If it is a very temporary problem though, this is often a nice                                                                                                                               
        compromise:                                                                                                                                                                                           
                                                                                                                                                                                                              
            yum-config-manager --save --setopt=rhel-fast-datapath.skip_if_unavailable=true                                                                                                                    
                                                                                                                                                                                                              
failure: repodata/repomd.xml from rhel-fast-datapath: [Errno 256] No more mirrors to try.                                                                                                                     
http://base-4-0.ocp.svc/rhel-fast-datapath/repodata/repomd.xml: [Errno 14] curl#6 - "Could not resolve host: base-4-0.ocp.svc; Unknown error"                                                                 
Loaded plugins: ovl, product-id, search-disabled-repos, subscription-manager                                                                                                                                  
This system is not registered with an entitlement server. You can use subscription-manager to register.                                                                                                       
http://base-4-0.ocp.svc/rhel-fast-datapath/repodata/repomd.xml: [Errno 14] curl#6 - "Could not resolve host: base-4-0.ocp.svc; Unknown error"                                                                 
Trying other mirror.                                                                                                                                                                                          
                                                                                                                                                                                                              
                                                                                                                                                                                                              
 One of the configured repositories failed (rhel-fast-datapath),                                                                                                                                              
 and yum doesn't have enough cached data to continue. At this point the only                                                                                                                                  
 safe thing yum can do is fail. There are a few ways to work "fix" this:                                                                                                                                      
                                                                                                                                                                                                              
     1. Contact the upstream for the repository and get them to fix the problem.                                                                                                                              
                                                                                                                                                                                                              
     2. Reconfigure the baseurl/etc. for the repository, to point to a working                                                                                                                                
        upstream. This is most often useful if you are using a newer                                                                                                                                          
        distribution release than is supported by the repository (and the                                                                                                                                     
        packages for the previous distribution release still work).                                                                                                                                           
                                                                                                                                                                                                              
     3. Run the command with the repository temporarily disabled                                                                                                                                              
            yum --disablerepo=rhel-fast-datapath ...                                                                                                                                                          
                                                                                                                                                                                                              
     4. Disable the repository permanently, so yum won't use it by default. Yum                                                                                                                               
        will then just ignore the repository until you permanently enable it                                                                                                                                  
        again or use --enablerepo for temporary usage:                                                                                                                                                        
                                                                                                                                                                                                              
            yum-config-manager --disable rhel-fast-datapath                                                                                                                                                   
        or                                                                                                                                                                                                    
            subscription-manager repos --disable=rhel-fast-datapath                                                                                                                                           

     5. Configure the failing repository to be skipped, if it is unavailable.
        Note that yum will try to contact the repo. when it runs most commands,
        so will have to try and fail each time (and thus. yum will be be much
        slower). If it is a very temporary problem though, this is often a nice
        compromise:

            yum-config-manager --save --setopt=rhel-fast-datapath.skip_if_unavailable=true

failure: repodata/repomd.xml from rhel-fast-datapath: [Errno 256] No more mirrors to try.
http://base-4-0.ocp.svc/rhel-fast-datapath/repodata/repomd.xml: [Errno 14] curl#6 - "Could not resolve host: base-4-0.ocp.svc; Unknown error"
Loaded plugins: ovl, product-id, search-disabled-repos, subscription-manager
This system is not registered with an entitlement server. You can use subscription-manager to register.
Cleaning repos: rhel-fast-datapath rhel-server rhel-server-extras
              : rhel-server-optional rhel-server-ose rhel-server-rhscl
Removing intermediate container 10d6f84d73aa

```

[the remark this answers is not in the document]
Comment (pliurh) [reply Dockerfile.sriov-network-config-daemon]: Dockerfile.rhel7 and Dockerfile.sriov-network-config-daemon.rhel7 will be consumed by openshift CI. You don't need to build image with them.

File: pkg/daemon/daemon.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (pliurh) [reply pkg/daemon/daemon.go]: For now, I prefer to only skip that for Mellanox's nic.

File: pkg/utils/utils.go [hunk 17]
```diff
@@ -131,6 +131,9 @@ func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState) error {
 }
 
 func needUpdate(iface *sriovnetworkv1.Interface, ifaceStatus *sriovnetworkv1.InterfaceExt) bool {
+	if ifaceStatus.Vendor == "15b3" && ifaceStatus.TotalVfs < iface.NumVfs {
```
Comment (moshe010) [inline pkg/utils/utils.go]: why do we need this change?

Comment (Mmduh-483) [reply pkg/utils/utils.go]: if the new numOfVfs > old totalVfs then the generic plugin will fail

File: pkg/plugins/mellanox/mellanox_plugin.go [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+			attrs = append(attrs, "LINK_TYPE_P2")
+			fwData.firstPort = false
+		}
+	}
+	mstData, err := mstconfigReadData(pciAddress, attrs...)
+	if err != nil {
+		return nil, err
+	}
+	result := map[string]string{}
+	mstData = strings.Trim(mstData, " \n")
```
Comment (moshe010) [inline pkg/plugins/mellanox/mellanox_plugin.go]: remove space

**Comment by pliurh** [comment]
@Mmduh-483 please do rebase before we can merge it.

**Comment by Mmduh-483** [comment]
@pliurh I rebased it
